Docs: Reorganize Configuration Documentation

khawarlatifkhan opened the pull request:
<!--
    Thank you for contributing!

    ESLint adheres to the [JS Foundation Code of Conduct](https://js.foundation/community/code-of-conduct).
-->

#### Prerequisites checklist

- [ X ] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/master/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

[ X ] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/master/templates/rule-change-proposal.md))
[ ] Add autofixing to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

<!--
    If the item you've checked above has a template, please paste the template questions below and answer them. (If this pull request is addressing an issue, you can just paste a link to the issue here instead.)
-->

<!--
    Please ensure your pull request is ready:

    - Read the pull request guide (https://eslint.org/docs/developer-guide/contributing/pull-requests)
    - Include tests for this change
    - Update documentation for this change (if appropriate)
-->

<!--
    The following is required for all pull requests:
-->

#### What changes did you make? (Give an overview)
- Reorganized the ESLint configuration documentation by dividing the single document into 5 sections (viz. Configuration Files, Language Options, Rules, Plugins, and Ignoring Code).
- Created an index (README.md) file for the "Configuring ESLint" directory, to improve the structure of the documentation.
- Updated and added new links to the .md files where necessary.

#### Is there anything you'd like reviewers to focus on?
At the beginning of all 5 sections, a brief introduction/context is needed so that the readers get a little idea of what they're going to gain from this section. Some feedback on that will be of great help.

khawarlatifkhan commented:
@kaicataldo added all the files to "configuring-ESLint" directory here. For some reason, 1 of the check is failing and I can't seem to understand that.
Looking forward to your feedback.

kaicataldo commented:
Amazing work here! The current failing check is due to us also linting the markdown files and there being a few linting errors that need to be fixed up. You can see the results [here](https://github.com/eslint/eslint/pull/13837/checks?check_run_id=1389648537) or see the linting results locally by running `npm run lint` in your terminal.

@eslint/eslint-team A lot of work has gone into this first draft of the revamped configuration documentation, and feedback from the team would be useful at this point. I'll try to get to reviewing it myself ASAP.

khawarlatifkhan commented:
Thank you for the appreciation @kaicataldo. I'm still a little confused about the failing checks. Is there anything I can do to fix it?

kaicataldo commented:
Yes, there is!

Here are the reported errors:
```
docs/user-guide/configuring-ESLint/plugins.md: 63: MD038/no-space-in-code Spaces inside code span elements [Context: "` rule for the named code bloc..."]
docs/user-guide/configuring-ESLint/README.md: 1: MD022/blanks-around-headings Headings should be surrounded by blank lines [Expected: 1; Actual: 0; Below] [Context: "# Configuring ESLint"]
docs/user-guide/configuring-ESLint/README.md: 19: MD032/blanks-around-lists Lists should be surrounded by blank lines [Context: "* [Configuration File Formats]..."]
docs/user-guide/configuring-ESLint/README.md: 28: MD032/blanks-around-lists Lists should be surrounded by blank lines [Context: "* [Specifying Environments](./..."]
docs/user-guide/configuring-ESLint/README.md: 33: MD032/blanks-around-lists Lists should be surrounded by blank lines [Context: "* [Configuring Rules](./rules...."]
docs/user-guide/configuring-ESLint/README.md: 37: MD032/blanks-around-lists Lists should be surrounded by blank lines [Context: "* [Specifying Parser](./plugin..."]
docs/user-guide/configuring-ESLint/README.md: 42: MD032/blanks-around-lists Lists should be surrounded by blank lines [Context: "* [`ignorePatterns` in config ..."]
```

To break this down:

- `docs/user-guide/configuring-ESLint/README.md: 42` is the file and line number of the error. 
- `MD032/blanks-around-lists Lists should be surrounded by blank lines` is the linting error that needs to be corrected.

Does that make sense?

nzakas commented:
It looks like the directory name has some uppercase letters, can you please be sure to use all lowercase? With the exception of README files, we use lowercase for everything.

Also, is there a reason this isn’t just in the /configuring directory we already have?

I’ll try to look at the actual text next week.

khawarlatifkhan commented:
Thank you so much @kaicataldo. I updated all the files and removed the errors.

@nzakas, thank you for pointing the mistake out. I've changed the directory name to lowercase configuring-eslint.
As for the already-present "configuring" directory, we previously had a single file for configuration. With this directory, we've split the documentation into 5 sections (or separate .md files), thus dedicating a whole directory (folder) to configuration.

kaicataldo commented:
I was thinking we'd reuse the existing `configuring/` directory as well :+1:

nzakas commented:
Yeah, I think we need to keep the same directory for this effort. Otherwise, it will affect SEO. Plus, we want to be sure any existing links to the configuring page seamlessly go to the new page.

khawarlatifkhan commented:
Sounds good. I'll change the diretory name from `configuring-eslint/` to `configuring/`. That'll take care of it.

nzakas commented on docs/user-guide/configuring/configuration-files.md:
```diff
@@ -0,0 +1,394 @@
+
+## Extending Configuration Files
+
+A configuration file can extend the set of enabled rules from base configurations.
```
Can we make this a bit clearer? Configurations include not just rules, but plugins, processors, environments, globals, and more. It might be more accurate to say that a configuration file can inherit all of the options in another configuration file and then override and modify those options? (Might need some better wording, but I think you get the idea.)

khawarlatifkhan replied:
I think I understand what you've mentioned here. Below is a modified version, lmk if it works:

> A configuration file, once extended, can inherit all the traits of another configuration file (including rules, plugins, and language options) and modify all the options.

nzakas replied:
Excellent!

nzakas commented on docs/user-guide/configuring/configuration-files.md:
```diff
@@ -0,0 +1,394 @@
+The `extends` property value is either:
+
+* a string that specifies a configuration (either a path to a config file, the name of a shareable config, `eslint:recommended`, or `eslint:all`)
+* an array of strings: each additional configuration extends the preceding configurations
```
```suggestion
* an array of strings where each additional configuration extends the preceding configurations
```

nzakas commented on docs/user-guide/configuring/configuration-files.md:
```diff
@@ -0,0 +1,394 @@
+* a string that specifies a configuration (either a path to a config file, the name of a shareable config, `eslint:recommended`, or `eslint:all`)
+* an array of strings: each additional configuration extends the preceding configurations
+
+ESLint extends configurations recursively, so a base configuration can also have an `extends` property. Relative paths and shareable config names in an `extends` property are resolved from the location of the config file where they appear.
```
It seems like we are missing an important piece of information here: the way that config names are resolved. So `example` is resolved to `eslint-config-example`, for example.

nzakas commented on docs/user-guide/configuring/configuration-files.md:
```diff
@@ -0,0 +1,394 @@
+* enable additional rules
+* change an inherited rule's severity without changing its options:
+    * Base config: `"eqeqeq": ["error", "allow-null"]`
+    * Derived config: `"eqeqeq": "warn"`
```
We haven't introduced the term "derived config" anywhere, so I'm not sure what this means. It may be helpful to define each of these terms first to make the example clear.

nzakas replied:
All the terms are okay as long as we define them upfront.

nzakas commented on docs/user-guide/configuring/configuration-files.md:
```diff
@@ -0,0 +1,394 @@
+
+### Using eslint-recommended
+
+An `extends` property value `"eslint:recommended"` enables a subset of core rules that report common problems, which have a checkmark (recommended) on the [rules page](https://eslint.org/docs/rules/). The recommended subset can change only at major versions of ESLint.
```
```suggestion
Using `"eslint:recommended"` in the `extends` property enables a subset of core rules that report common problems (these rules are identified with a checkmark (recommended) on the [rules page](https://eslint.org/docs/rules/)). The `eslint:recommended` config only changes with major versions of ESLint.
```

nzakas commented on docs/user-guide/configuring/configuration-files.md:
```diff
@@ -0,0 +1,394 @@
+    * Derived config: `"quotes": ["error", "single"]`
+    * Resulting actual config: `"quotes": ["error", "single"]
+
+### Using eslint-recommended
```
```suggestion
### Using `eslint:recommended`
```

nzakas commented on docs/user-guide/configuring/configuration-files.md:
```diff
@@ -0,0 +1,394 @@
+
+An `extends` property value `"eslint:recommended"` enables a subset of core rules that report common problems, which have a checkmark (recommended) on the [rules page](https://eslint.org/docs/rules/). The recommended subset can change only at major versions of ESLint.
+
+If your configuration extends the recommended rules: after you upgrade to a newer major version of ESLint, review the reported problems before you use the `--fix` option on the [command line]((https://eslint.org/docs/user-guide/command-line-interface)#fix), so you know if a new fixable recommended rule will make changes to the code.
```
```suggestion
If your configuration extends the `eslint:recommended`, be sure to review the reported problems after upgrading to a new major version of ESLint before you use the `--fix` option on the [command line]((https://eslint.org/docs/user-guide/command-line-interface)#fix). Doing so ensures you'll be aware of the new changes ESLint might make to your code.
```

nzakas commented on docs/user-guide/configuring/configuration-files.md:
```diff
@@ -0,0 +1,394 @@
+
+If your configuration extends the recommended rules: after you upgrade to a newer major version of ESLint, review the reported problems before you use the `--fix` option on the [command line]((https://eslint.org/docs/user-guide/command-line-interface)#fix), so you know if a new fixable recommended rule will make changes to the code.
+
+The `eslint --init` command can create a configuration so you can extend the recommended rules.
```
This seems out of place here. Maybe we can just say:

```suggestion
Here's an example of extending `eslint:recommended` to override some of the default options:
```

nzakas commented on docs/user-guide/configuring/configuration-files.md:
```diff
@@ -0,0 +1,394 @@
+
+### Using a shareable configuration package
+
+A [sharable configuration](https://eslint.org/docs/developer-guide/shareable-configs) is an npm package that exports a configuration object. Make sure the package has been installed to a directory where ESLint can require it.
```
We probably need to be more specific about the location of the package. Maybe we can just say that you should install it in your project root?

nzakas commented on docs/user-guide/configuring/configuration-files.md:
```diff
@@ -0,0 +1,394 @@
+
+A [sharable configuration](https://eslint.org/docs/developer-guide/shareable-configs) is an npm package that exports a configuration object. Make sure the package has been installed to a directory where ESLint can require it.
+
+The `extends` property value can omit the `eslint-config-` prefix of the package name.
```
I think this section needs to be moved up to be closer to the start of the section on `extends`

nzakas commented on docs/user-guide/configuring/configuration-files.md:
```diff
@@ -0,0 +1,394 @@
+
+The `extends` property value can omit the `eslint-config-` prefix of the package name.
+
+The `eslint --init` command can create a configuration so you can extend a popular style guide (for example, `eslint-config-standard`).
```
Again, I don't think it's worth talking about `eslint --init` in this section. Everything that command does can be done by hand, so I'd suggest we focus on manual configuration edits here.

nzakas commented on docs/user-guide/configuring/configuration-files.md:
````diff
@@ -0,0 +1,394 @@
+  no-empty: warn
+```
+
+### Using the configuration from a plugin
````
```suggestion
### Using a configuration from a plugin
```

nzakas commented on docs/user-guide/configuring/configuration-files.md:
```diff
@@ -0,0 +1,394 @@
+
+### Using the configuration from a plugin
+
+A [plugin](https://eslint.org/docs/developer-guide/working-with-plugins) is an npm package that usually exports rules. Some plugins also export one or more named [configurations](https://eslint.org/docs/developer-guide/working-with-plugins#configs-in-plugins). Make sure the package has been installed in a directory where ESLint can require it.
```
Plugins aren't just for rules and are intended to provide all kinds of extensions to ESLint. So rather than saying plugins normally export rules and only sometimes export configs, maybe we can say that plugins export a number of things including rules and configs?

khawarlatifkhan replied:
Oh yes. This'll def make it much better. Will the following content work?

> A [plugin](https://eslint.org/docs/developer-guide/working-with-plugins) is an npm package that can add various functions to ESLint. A plugin can perform numerous functions, including but not limited to the export of rules and [configurations](https://eslint.org/docs/developer-guide/working-with-plugins#configs-in-plugins). Make sure the package has been installed in a directory where ESLint can require it.

nzakas commented on docs/user-guide/configuring/configuration-files.md:
```diff
@@ -0,0 +1,394 @@
+
+**Important:** This configuration is **not recommended for production use** because it changes with every minor and major version of ESLint. Use it at your own risk.
+
+If you configure ESLint to automatically enable new rules when you upgrade, ESLint can report new problems when there are no changes to source code, therefore any newer minor version of ESLint can behave as if it has breaking changes.
```
I'm not sure what this means. Do you mean if you use `eslint:all`?

nzakas commented on docs/user-guide/configuring/configuration-files.md:
```diff
@@ -0,0 +1,394 @@
+
+You might enable all core rules as a shortcut to explore rules and options while you decide on the configuration for a project, especially if you rarely override options or disable rules. The default options for rules are not endorsements by ESLint (for example, the default option for the [`quotes`](https://eslint.org/docs/rules/quotes) rule does not mean double quotes are better than single quot ...
+
+If your configuration extends all core rules: after you upgrade to a newer major or minor version of ESLint, review the reported problems before you use the `--fix` option on the [command line](https://eslint.org/docs/user-guide/command-line-interface#fix), so you know if a new fixable rule will make changes to the code.
```
```suggestion
If your configuration extends `eslint:all`, after you upgrade to a newer major or minor version of ESLint, review the reported problems before you use the `--fix` option on the [command line](https://eslint.org/docs/user-guide/command-line-interface#fix), so you know if a new fixable rule will make changes to the code.
```

nzakas commented on docs/user-guide/configuring/configuration-files.md:
```diff
@@ -0,0 +1,394 @@
+
+<b>v4.1.0+.</b> Sometimes a more fine-controlled configuration is necessary, for example, if the configuration for files within the same directory has to be different. Therefore you can provide configurations under the `overrides` key that will only apply to files that match specific glob patterns, using the same format you would pass on the command line (e.g., `app/**/*.test.js`).
+
+### How does it work
```
```suggestion
### How overrides work
```

nzakas commented on docs/user-guide/configuring/configuration-files.md:
```diff
@@ -0,0 +1,394 @@
+
+### How does it work
+
+* The patterns are applied against the file path relative to the directory of the config file. For example, if your config file has the path `/Users/john/workspace/any-project/.eslintrc.js` and the file you want to lint has the path `/Users/john/workspace/any-project/lib/util.js`, then the pattern provided in `.eslintrc.js` will be executed against the relative path `lib/util.js`.
```
Before going into how overrides work, I think it would be helpful to show an example.

khawarlatifkhan replied:
Do you think that moving the `### Example configuration` before `### How do overrides work?`, will take care of this issue?

nzakas replied:
Yes. I think I'd start with saying something like, "It's possible to override settings based on file glob patterns in your configuration by using the `overrides` key, like this:"

Then have the example and follow that with the content from this section.

khawarlatifkhan replied:
Awesome, thanks. Readjusting the content of this section then.

nzakas commented on docs/user-guide/configuring/configuration-files.md:
```diff
@@ -0,0 +1,394 @@
+
+`~/` refers to [the home directory of the current user on your preferred operating system](https://nodejs.org/api/os.html#os_os_homedir). The personal configuration file being referred to here is `~/.eslintrc.*` file, which is currently handled differently than other configuration files.
+
+### How does ESLint find personal configuration files
```
```suggestion
### How does ESLint find personal configuration files?
```

nzakas commented on docs/user-guide/configuring/configuration-files.md:
```diff
@@ -0,0 +1,394 @@
+
+If `eslint` could find configuration files in the project, `eslint` ignores `~/.eslintrc.*` file even if it's in an ancestor directory of the project directory.
+
+### How do personal configuration files behave
```
```suggestion
### How do personal configuration files behave?
```

nzakas reviewed (commented):
Add a bunch of comments for the first couple of pages. Need to take a break and come back to the others later.

mdjermanovic commented on docs/user-guide/configuring/configuration-files.md:
```diff
@@ -0,0 +1,394 @@
+ESLint supports configuration files in several formats:
+
+* **JavaScript** - use `.eslintrc.js` and export an object containing your configuration.
+* **JavaScript (ESM)** - use `.eslintrc.cjs` when running ESLint in JavaScript packages that specify `"type":"module"` in their `package.json`. Note that ESLint does not support ESM configuration at this time.
```
This was added per https://github.com/eslint/rfcs/pull/43. It's basically the only way to use ESLint in `"type": "module"` packages at the moment.

khawarlatifkhan replied:
Does this mean we're keeping it as it is rn?
@mdjermanovic @kaicataldo

mdjermanovic replied:
I think we should keep this.

mdjermanovic commented on docs/user-guide/configuring/configuration-files.md:
```diff
@@ -0,0 +1,394 @@
+If there are multiple configuration files in the same directory, ESLint will only use one. The priority order is as follows:
+
+1. `.eslintrc.js`
+1. `.eslintrc.cjs`
```
This was implemented in https://github.com/eslint/eslint/pull/12321, per https://github.com/eslint/rfcs/pull/43.

khawarlatifkhan replied:
Does this mean we're keeping it as it is rn?
@mdjermanovic @kaicataldo

mdjermanovic replied:
> Does this mean we're keeping it as it is rn?

Yes, it works.

khawarlatifkhan commented on docs/user-guide/configuring/configuration-files.md:
```diff
@@ -0,0 +1,394 @@
+
+There are two ways to use configuration files.
+
+The first way to use configuration files is via `.eslintrc.*` and `package.json` files. ESLint will automatically look for them in the directory of the file to be linted, and in successive parent directories all the way up to the root directory of the filesystem (unless `root: true` is specified). This option is useful when you want different configurations for different parts of a project or whe ...
```
"This option" refers to the use of `.eslintrc.*` and `package.json` files.
Is this technically correct though? If it is, I can edit the sentence something like this:

`.eslintrc.*` and `package.json` files can be useful when you want different configurations for different parts of a project or...

nzakas replied:
I think maybe just saying "Configuration files can be useful..." would be clear enough.

khawarlatifkhan commented on docs/user-guide/configuring/configuration-files.md:
```diff
@@ -0,0 +1,394 @@
+
+If you are using one configuration file and want ESLint to ignore any `.eslintrc.*` files, make sure to use `--no-eslintrc` along with the `-c` flag.
+
+In each case, the settings in the configuration file override default settings.
```
As per my understandings, it refers to all the default settings for configuration files. Whichever way a configuration file is used, it is going to override the default settings already available to the users.

Please correct me if I'm wrong.

khawarlatifkhan replied:
Got it. Will remove the "default settings" from all the instances in this file.

khawarlatifkhan commented on docs/user-guide/configuring/configuration-files.md:
```diff
@@ -0,0 +1,394 @@
+
+### Comments in configuration files
+
+Both the JSON and YAML configuration file formats support comments (`package.json` files should not include them). You can use JavaScript-style comments or YAML-style comments in either type of file and ESLint will safely ignore them. This allows your configuration files to be more human-friendly. For example:
```
Thank you for pointing this error out. This is something I wasn't aware of. If I edit it in the following way, will that work?

> Both the JSON and YAML configuration file formats support comments (`package.json` files should not include them). You can use JavaScript-style comments for JSON files and YAML-style comments for YAML files. On detecting comments in a file, ESLint will safely ignore them. This allows your configuration files to be more human-friendly.

Also, do you think we need to add one example each for JavaScript-style and YAML-style comments?

khawarlatifkhan replied:
Awesome. Does the following content look good then? Don't want to make any technical errors, so I thought I'd get it reviewed first.

> Both the JSON and YAML configuration file formats support comments (package.json files should not include them). You can use JavaScript-style comments for JSON files and YAML-style comments for YAML files. On detecting comments in a file, ESLint will safely ignore them. This allows your configuration files to be more human-friendly.
> For JavaScript-style comments:

```js
{
    "env": {
        "browser": true
    },
    "rules": {
        // Override our default settings just for this directory
        "eqeqeq": "warn",
        "strict": "off"
    }
}
```

> For YAML-style comments:

```js
{
    "env": {
        "browser": true
    },
    "rules": {
        # Override our default settings just for this directory
        "eqeqeq": "warn",
        "strict": "off"
    }
}
```

nzakas replied:
> On detecting comments in a file, ESLint will safely ignore them. 

Maybe, "ESLint safely ignores comments in configuration files."?

Also, the second example still looks like a JSON file. I'd suggest making it look more like YAML:

```yaml
env:
    browser: true
rules:
    # Override default settings
    eqeqeq: warn
    strict: off
```

kaicataldo reviewed (commented):
Awesome work on this! I have some suggestions/comments. I'm currently about halfway through and will try to finish it up over the next day or so.

khawarlatifkhan commented on docs/user-guide/configuring/configuration-files.md:
```diff
@@ -0,0 +1,393 @@
+    1. `-c`, `--config`
+1. Project-level configuration:
+    1. `.eslintrc.*` or `package.json` file in the same directory as linted file
+    1. Continue searching for `.eslintrc` and `package.json` files in ancestor directories (parent has the highest precedence, then grandparent, etc.), up to and including the root directory or until a config with `"root": true` is found.
```
Oh yes. It's pretty clear without the text in the parantheses.

khawarlatifkhan commented on docs/user-guide/configuring/configuration-files.md:
```diff
@@ -0,0 +1,393 @@
+
+### Using `eslint:recommended`
+
+Using `"eslint:recommended"` in the `extends` property enables a subset of core rules that report common problems (these rules are identified with a checkmark (recommended) on the [rules page](https://eslint.org/docs/rules/)). The `eslint:recommended` config only changes with major versions of ESLint.
```
Done. Thank you!!!

khawarlatifkhan commented on docs/user-guide/configuring/configuration-files.md:
```diff
@@ -0,0 +1,393 @@
+
+Using `"eslint:recommended"` in the `extends` property enables a subset of core rules that report common problems (these rules are identified with a checkmark (recommended) on the [rules page](https://eslint.org/docs/rules/)). The `eslint:recommended` config only changes with major versions of ESLint.
+
+If your configuration extends the `eslint:recommended`, be sure to review the reported problems after upgrading to a new major version of ESLint before you use the `--fix` option on the [command line]((https://eslint.org/docs/user-guide/command-line-interface)#fix). Doing so ensures you'll be aware of the new changes ESLint might make to your code.
```
So, we can simply remove this paragraph as it is a clearly-understood practice? Or do we rephrase it in a way that it becomes more generic? What do you suggest?

kaicataldo replied:
I personally think it's fine to just remove it. @nzakas thoughts?

nzakas replied:
I'm fine with removing, too.

khawarlatifkhan commented on docs/user-guide/configuring/configuration-files.md:
```diff
@@ -0,0 +1,393 @@
+
+If your configuration extends the `eslint:recommended`, be sure to review the reported problems after upgrading to a new major version of ESLint before you use the `--fix` option on the [command line]((https://eslint.org/docs/user-guide/command-line-interface)#fix). Doing so ensures you'll be aware of the new changes ESLint might make to your code.
+
+Here's an example of extending `eslint:recommended` to override some of the default options:
```
This seems good! Thank you.

khawarlatifkhan commented on docs/user-guide/configuring/configuration-files.md:
```diff
@@ -0,0 +1,393 @@
+        "semi": ["error", "always"],
+
+        // override default options for rules from base configurations
+        "comma-dangle": ["error", "always"],
```
Can we use `for-direction` or `no-empty` instead?

Also, can you help me a little with those? I'm not so sure how to actually use these codes in a file, so I might make a mistake.

kaicataldo replied:
`for-direction` sounds good! we should be able to replace this line with `"for-direction": "off",` :)

nzakas commented on docs/user-guide/configuring/ignoring-code.md:
```diff
@@ -0,0 +1,182 @@
+
+## `ignorePatterns` in config files
+
+You can tell ESLint to ignore specific files and directories by `ignorePatterns` in your config files. Each value of `ignorePatterns` is the same pattern as each line of `.eslintignore` in the next section.
```
```suggestion
You can tell ESLint to ignore specific files and directories by `ignorePatterns` in your config files. Each value of `ignorePatterns` is a glob pattern similar to what you would use on the command line.
```

I don't think referring to `.eslintignore` helps here, so I'd suggest removing it.

nzakas commented on docs/user-guide/configuring/ignoring-code.md:
````diff
@@ -0,0 +1,182 @@
+}
+```
+
+* The `ignorePatterns` property affects only the directory that the config file is placed in.
````
I think maybe you mean this?

```suggestion
* Glob patterns in `ignorePatterns` are relative to the directory that the config file is placed in.
```

nzakas commented on docs/user-guide/configuring/ignoring-code.md:
```diff
@@ -0,0 +1,182 @@
+
+* The `ignorePatterns` property affects only the directory that the config file is placed in.
+* You cannot write `ignorePatterns` property under `overrides` property.
+* `.eslintignore` can override the `ignorePatterns` property of config files.
```
I'm not sure if "override" is the correct word here. Do you mean if a `.eslintignore` file negates an ignore then it will override what's in the config file?

kaicataldo replied:
Yes, I believe that's the intention here (e.g. using `.eslintignore` to unignore a specific file that's ignored by a glob pattern in `ignorePatterns`). Maybe something along the lines of the following would be clearer?

```suggestion
* Patterns defined in `.eslintignore` take precedence over the `ignorePatterns` property of config files.
```

mysticatea replied:
Actually, it's just combined. It's similar to:

```js
const patterns = [
    ...config.ignorePatterns,
    ...fs.readFileSync(".eslintignore", "utf8").split("\n")
]
```

nzakas commented on docs/user-guide/configuring/ignoring-code.md:
```diff
@@ -0,0 +1,182 @@
+
+If a config is provided via the `--config` CLI option, the ignore patterns that start with `/` in the config are relative to the current working directory rather than the base directory of the given config. For example, if `--config configs/.eslintrc.json` is present, the ignore patterns in the config are relative to `.` rather than `./configs`.
+
+## `.eslintignore`
```
```suggestion
## The `.eslintignore` file
```

nzakas commented on docs/user-guide/configuring/ignoring-code.md:
```diff
@@ -0,0 +1,182 @@
+
+There are also some exceptions to these rules:
+
+* If the path to lint is a glob pattern or directory path and contains a Dotfolder, all Dotfiles and Dotfolders will be linted. This includes sub-dotfiles and sub-dotfolders that are buried deeper in the directory structure.
```
Dotfolder and Dotfile should probably be dot-folder and dot-file.

khawarlatifkhan replied:
we can hyphenate these two. A small question here, though: what about the `sub-dotfiles` and `sub-dotfolders`?

nzakas replied:
I think we can probably replace those with just dot-files and dot-folders. I'm not sure people will understand sub-dotfiles and sub-dotfolders.

nzakas commented on docs/user-guide/configuring/ignoring-code.md:
```diff
@@ -0,0 +1,182 @@
+
+  For example, `eslint .config/` will lint all Dotfolders and Dotfiles in the `.config` directory, including immediate children as well as children that are deeper in the directory structure.
+
+* If the path to lint is a specific file path and the `--no-ignore` flag has been passed, ESLint will lint the file regardless of the implicit ignore rules.
```
We may need to reformat this section. The asterisk will create a one-item bullet list here and the next paragraph will not be indented.

khawarlatifkhan replied:
<img width="1038" alt="Screenshot 2020-11-28 at 12 05 31 PM" src="https://user-images.githubusercontent.com/66240534/100495956-08e21880-3172-11eb-920b-4d9230d49c81.png">

This is how it appears in the current state.  The examples under the bulleted items are indented. Isn't that how we want it to be? Or shall I remove the indents?

nzakas replied:
That actually looks perfect, thank you!

nzakas commented on docs/user-guide/configuring/ignoring-code.md:
```diff
@@ -0,0 +1,182 @@
+
+This message occurs because, normally, this file would be ignored by ESLint's implicit ignore rules (as mentioned above). A negated ignore rule in your `.eslintignore` file would override the implicit rule and reinclude this file for linting. Additionally, in this specific case, `--no-ignore` could be used to lint the file as well.
+
+## Disabling Inline Comments
```
This seems a bit out of place to me as it doesn't relate to ignoring files.

khawarlatifkhan replied:
Can we move this entire section under `### Comments in Configuration Files` heading in the `Configuration Files` section? Here's how it will look:

# Configuration Files
## Using Configuration Files
### Comments in Configuration Files
#### Disabling Inline Comments
##### Report unused `eslint-disable` comments

khawarlatifkhan replied:
Alternatively, as @kaicataldo suggested in one of our conversations, we can move this particular heading to the `Rules` section. This can go under the `Disabling Rules` heading.

nzakas replied:
I think moving it under the Rules section makes sense.

nzakas commented on docs/user-guide/configuring/ignoring-code.md:
```diff
@@ -0,0 +1,182 @@
+
+This setting is similar to [--no-inline-config](https://eslint.org/docs/user-guide/command-line-interface#--no-inline-config) CLI option.
+
+### Report unused `eslint-disable` comments
```
Same here. There may be a better place to put these two sections.

nzakas commented on docs/user-guide/configuring/language-options.md:
```diff
@@ -0,0 +1,214 @@
+
+## Specifying Environments
+
+An environment defines global variables that are predefined. The available environments are:
```
```suggestion
An environment provides predefined global variables. The available environments are:
```

nzakas commented on docs/user-guide/configuring/language-options.md:
```diff
@@ -0,0 +1,214 @@
+
+## Specifying Globals
+
+The [no-undef](https://eslint.org/docs/rules/no-undef) rule will warn on variables that are accessed but not defined within the same file. If you are using global variables inside of a file then it's worthwhile to define those globals so that ESLint will not warn about their usage. You can define global variables either using comments inside of a file or in the configuration file.
```
There are a bunch of rules that rely on global variable information, so I wouldn't mention any specific rules here (it will be easy to forget to update this if things change.)

khawarlatifkhan replied:
I think I understand what you're saying here. To make the documentation reusable, it'll def be a good idea to stay away from specific rules here. I have a couple of suggestions on how to improve this paragraph, lmk what you think works well:

> ESLint has [rules] (https://eslint.org/docs/rules) that will warn if so some variables are accessed but not defined within the same file. If you are using global variables inside of a file then it's worthwhile to define those globals so that ESLint will not warn about their usage. You can define global variables either using comments inside of a file or in the configuration file.

> If you are using global variables inside of a file then it's worthwhile to define those globals so that ESLint will not warn about their usage. You can define global variables either using comments inside of a file or in the configuration file.

kaicataldo replied:
My suggestion:

```suggestion
Some of ESLint's core rules rely on knowledge of the global variables available to your code at runtime. Since these can vary greatly between different environments as well as be modified at runtime, ESLint makes no assumptions about what global variables exist in your execution environment. If you would like to use rules that require knowledge of what global variables are available, you can define global variables in your configuration file or by using configure comments in your source code.
```

mysticatea replied:
I like mentioning to the related rules or `eslint:recommended`.

If someone configured globals, but ESLint doesn't warn about globals, they would be surprised. ESLint has many rules and it's hard to find related rules, so it's worthful that the "Specifying Globals" section describes how we configure warnings about globals. Probably `eslint:recommended` is suitable.

kaicataldo replied:
We have a few different suggestions from team members here. I agree that it makes sense to avoid mentioning specific rules here for maintenance reasons, and don't feel strongly about how we go about solving that. Could we provide @khawarlatifkhan with a clear path forward so that we can land this?

nzakas replied:
In the interest of moving things forward, let's go with @kaicataldo's suggestion here and we can always tweak the docs after it's merged.

nzakas commented on docs/user-guide/configuring/language-options.md:
```diff
@@ -0,0 +1,214 @@
+
+For historical reasons, the boolean value `false` and the string value `"readable"` are equivalent to `"readonly"`. Similarly, the boolean value `true` and the string value `"writeable"` are equivalent to `"writable"`. However, the use of older values is deprecated.
+
+**Note:** Enable the [no-global-assign](https://eslint.org/docs/rules/no-global-assign) rule to disallow modifications to read-only global variables in your code.
```
Again, not sure we want to reference specific rules here in case things change. We can always reference from the rule docs to this page.

khawarlatifkhan replied:
Please let me know if one of the following changes works in this case:

Suggestion one: deleting the note.

Suggestion two:

> **Note:** You can enable some ESLint [rules] (https://eslint.org/docs/rules) to disallow modifications to read-only global variables in your code.

kaicataldo replied:
Friendly ping @nzakas

nzakas replied:
```suggestion
```

nzakas replied:
I think it makes sense to remove it.

nzakas commented on docs/user-guide/configuring/plugins.md:
```diff
@@ -0,0 +1,177 @@
+
+By default, ESLint uses [Espree](https://github.com/eslint/espree) as its parser. You can optionally specify that a different parser should be used in your configuration file so long as the parser meets the following requirements:
+
+1. It must be a Node module loadable from the config file where it appears. Usually, this means you should install the parser package separately using npm.
```
```suggestion
1. It must be a Node module loadable from the config file where the parser is used. Usually, this means you should install the parser package separately using npm.
```

nzakas commented on docs/user-guide/configuring/rules.md:
```diff
@@ -0,0 +1,238 @@
+
+## Configuring Rules
+
+ESLint comes with a large number of rules. You can modify which rules your project uses either using configuration comments or configuration files. To change a rule setting, you must set the rule ID equal to one of these values:
```
```suggestion
ESLint comes with a large number of built-in rules and you can add more rules through plugins. You can modify which rules your project uses either using configuration comments or configuration files. To change a rule setting, you must set the rule ID equal to one of these values:
```

khawarlatifkhan commented on docs/user-guide/configuring/configuration-files.md:
```diff
@@ -0,0 +1,404 @@
+        "semi": ["error", "always"],
+
+        // override configuration set by extending "eslint:recommended"
+        "for-direction": "off",
```
Thank you!!
I was looking for a different example but couldn't find one. I was worried I might make a technical error. This is awesome.

khawarlatifkhan commented on docs/user-guide/configuring/configuration-files.md:
```diff
@@ -0,0 +1,404 @@
+
+The first way to use configuration files is via `.eslintrc.*` and `package.json` files. ESLint will automatically look for them in the directory of the file to be linted, and in successive parent directories all the way up to the root directory of the filesystem (unless `root: true` is specified). Configuration files can be useful when you want different configurations for different parts of a pr ...
+
+The second way to use configuration files is to save the file wherever you would like and pass its location to the CLI using the `-c` option, such as:
```
Oh this makes sense, yes.
I can change the `-c` to `--config`. Is that what you're suggesting? I'm sorry if I got it wrong.

btmills replied:
Yep, that’s precisely it!

khawarlatifkhan commented on docs/user-guide/configuring/configuration-files.md:
````diff
@@ -0,0 +1,393 @@
+  └── test.js
+```
+
+The configuration cascade works by using the closest `.eslintrc` file to the file being linted as the highest priority, then any configuration files in the parent directory, and so on. When you run ESLint on this project, all files in `lib/` will use the `.eslintrc` file at the root of the project as their configuration. When ESLint traverses into the `tests/` directory, it will then use `your-pr ...
````
Does something like the following work?

> The configuration cascade works by using the project-wide `eslintrc` file to the file being linted then any configuration files in the parent directory, and so on.

kaicataldo replied:
I don't think that suggestion is specific enough, but we're on the right track. The following is too wordy, but maybe someone else can find a way to pare it down for me?

> ESLint allows for configuration files in subdirectories to override values in the project-wide configuration when linting files in that subdirectory. Please note that only the configuration options defined in the subdirectories will be overridden. More specifically, ESLint starts from the  root configuration (the configuration file with `root: true` defined, usually in the root directory of the project), and if it finds a configuration file in a subdirectory, it will merge those values into the root configuration when linting the files in that subdirectory.

khawarlatifkhan replied:
@nzakas This is the only issue that needs to be resolved.

nzakas replied:
Yeah, I think we can make this a bit clearer. Here's my take:

> The configuration cascade works based on the location of the file being linted. If there is a `.eslintrc` file in the same directory as the file being linted, then that configuration takes precedence. ESLint then searches up the directory structure, merging any `.eslintrc` files it finds along the way until reaching either a `.eslintrc` file with `root: true` or the root directory.

mysticatea commented on docs/user-guide/configuring/ignoring-code.md:
```diff
@@ -0,0 +1,182 @@
+
+## `ignorePatterns` in Config Files
+
+You can tell ESLint to ignore specific files and directories using `ignorePatterns` in your config files. Each value of `ignorePatterns` is a glob pattern similar to what you would use on the command line.
```
I think this is a wrong explanation because the patterns in `ignorePatterns` are different from the glob patterns on the command line. The patterns are intended to be the same as `.gitignore`-style patterns, as same as `.eslintignore`.

Therefore, what the `The `.eslintignore` File` section described matches the `ignorePatterns` property. For example, it supports negative patterns (e.g., `!foo.js`).

Therefore, for example, `fs.readFileSync(".gitignore", "utf8").split("\n")` is a valid value of the `ignorePatterns` property.

kaicataldo replied:
Good catch. @khawarlatifkhan Do you need any additional information to fix this? There's a lot of implicit knowledge here :)

khawarlatifkhan replied:
I think I'll need some help with this so that I don't get anything wrong while editing.

kaicataldo replied:
Here's a suggestion:

```suggestion
You can tell ESLint to ignore specific files and directories using `ignorePatterns` in your config files. `ignorePatterns` patterns follow the same rules as `.eslintignore`. Please see the [the `.eslintignore` file documentation](./ignoring-code.md#the-eslintignore-file) to learn more.
```

kaicataldo commented:
Looks like we have a few outstanding comments, but this is looking really great. I'd love to get this to a place where we all feel good about landing it. Now that Season of Docs is finished, I want to make sure we're respectful of @khawarlatifkhan's time (though we would love it if you decided to continue contributing beyond this project!).

khawarlatifkhan commented:
> Looks like we have a few outstanding comments, but this is looking really great. I'd love to get this to a place where we all feel good about landing it. Now that Season of Docs is finished, I want to make sure we're respectful of @khawarlatifkhan's time (though we would love it if you decided to continue contributing beyond this project!).

Thank you so much @kaicataldo. I'd love to be an active contributor in the future as well. This has been a great experience.

kaicataldo commented:
> Thank you so much @kaicataldo. I'd love to be an active contributor in the future as well. This has been a great experience.

That would be wonderful! Please feel free to make PRs whenever you'd like!

nzakas reviewed (commented):
Sorry for the delay, my health hasn't been great the past couple of weeks so I'm just getting back around to ESLint stuff.

Overall, I think we are in good shape. I merged some of the suggestions that were already there to help us move forward.

@khawarlatifkhan I do want to be respectful of your time. Please let us know If you feel you can address the remaining issues (I think there's just a handful).

khawarlatifkhan commented:
> Sorry for the delay, my health hasn't been great the past couple of weeks so I'm just getting back around to ESLint stuff.
> 
> Overall, I think we are in good shape. I merged some of the suggestions that were already there to help us move forward.
> 
> @khawarlatifkhan I do want to be respectful of your time. Please let us know If you feel you can address the remaining issues (I think there's just a handful).

I hope you're feeling better now.
Thank you for taking care of a lot of things in this PR.

I'd love to work on all the remaining issues here. There are still a few comments/suggestions where your input would be of great help.

Looking forward.

nzakas commented on docs/user-guide/configuring/ignoring-code.md:
```diff
@@ -61,21 +60,21 @@ Please see [`.gitignore`](https://git-scm.com/docs/gitignore)'s specification fo
 
 * `node_modules/` is ignored.
-* Dotfiles (except for `.eslintrc.*`), as well as Dotfolders and their contents, are ignored.
+* Dot-files (except for `.eslintrc.*`), as well as Dot-folders and their contents, are ignored.
```
```suggestion
* dot-files (except for `.eslintrc.*`), as well as dot-folders and their contents, are ignored.
```

nzakas commented on docs/user-guide/configuring/ignoring-code.md:
```diff
@@ -61,21 +60,21 @@ Please see [`.gitignore`](https://git-scm.com/docs/gitignore)'s specification fo
 There are also some exceptions to these rules:
 
-* If the path to lint is a glob pattern or directory path and contains a Dotfolder, all Dotfiles and Dotfolders will be linted. This includes sub-dotfiles and sub-dotfolders that are buried deeper in the directory structure.
+* If the path to lint is a glob pattern or directory path and contains a Dot-folder, all Dot-files and Dot-folders will be linted. This includes dot-files and dot-folders that are buried deeper in the directory structure.
```
```suggestion
* If the path to lint is a glob pattern or directory path and contains a dot-folder, all dot-files and dot-folders will be linted. This includes dot-files and dot-folders that are buried deeper in the directory structure.
```

nzakas commented on docs/user-guide/configuring/ignoring-code.md:
```diff
@@ -61,21 +60,21 @@ Please see [`.gitignore`](https://git-scm.com/docs/gitignore)'s specification fo
+* If the path to lint is a glob pattern or directory path and contains a Dot-folder, all Dot-files and Dot-folders will be linted. This includes dot-files and dot-folders that are buried deeper in the directory structure.
 
-  For example, `eslint .config/` will lint all Dotfolders and Dotfiles in the `.config` directory, including immediate children as well as children that are deeper in the directory structure.
+  For example, `eslint .config/` will lint all Dot-folders and Dot-files in the `.config` directory, including immediate children as well as children that are deeper in the directory structure.
```
```suggestion
  For example, `eslint .config/` will lint all dot-folders and dot-files in the `.config` directory, including immediate children as well as children that are deeper in the directory structure.
```

nzakas commented on docs/user-guide/configuring/ignoring-code.md:
```diff
@@ -61,21 +60,21 @@ Please see [`.gitignore`](https://git-scm.com/docs/gitignore)'s specification fo
 * Allowlist and denylist rules specified via `--ignore-pattern` or `.eslintignore` are prioritized above implicit ignore rules.
 
-  For example, in this scenario, `.build/test.js` is the desired file to allowlist. Because all Dotfolders and their children are ignored by default, `.build` must first be allowlisted so that eslint becomes aware of its children. Then, `.build/test.js` must be explicitly allowlisted, while the rest of the content is denylisted. This is done with the following `.eslintignore` file:
+  For example, in this scenario, `.build/test.js` is the desired file to allowlist. Because all Dot-folders and their children are ignored by default, `.build` must first be allowlisted so that eslint becomes aware of its children. Then, `.build/test.js` must be explicitly allowlisted, while the rest of the content is denylisted. This is done with the following `.eslintignore` file:
```
```suggestion
  For example, in this scenario, `.build/test.js` is the desired file to allowlist. Because all dot-folders and their children are ignored by default, `.build` must first be allowlisted so that eslint becomes aware of its children. Then, `.build/test.js` must be explicitly allowlisted, while the rest of the content is denylisted. This is done with the following `.eslintignore` file:
```

nzakas commented on docs/user-guide/configuring/ignoring-code.md:
```diff
@@ -140,7 +139,7 @@ foo.js
 This message occurs because ESLint is unsure if you wanted to actually lint the file or not. As the message indicates, you can use `--no-ignore` to omit using the ignore rules.
 
-Consider another scenario where you may want to run ESLint on a specific Dotfile or Dotfolder, but have forgotten to specifically allow those files in your `.eslintignore` file. You would run something like this:
+Consider another scenario where you may want to run ESLint on a specific Dot-file or Dot-folder, but have forgotten to specifically allow those files in your `.eslintignore` file. You would run something like this:
```
```suggestion
Consider another scenario where you may want to run ESLint on a specific dot-file or dot-folder, but have forgotten to specifically allow those files in your `.eslintignore` file. You would run something like this:
```

nzakas reviewed (commented):
Thanks. I just added some suggestions for changing Dot-files to dot-files.

nzakas commented:
@khawarlatifkhan thanks so much. Are there other specific things you're waiting on feedback for right now? (It's easy to miss things when a PR is this large, so please let me know.)

khawarlatifkhan commented:
> @khawarlatifkhan thanks so much. Are there other specific things you're waiting on feedback for right now? (It's easy to miss things when a PR is this large, so please let me know.)

Thank you. There's only one comment left, where your input will be of help. I'm going to mention you in a comment there.

nzakas commented:
@khawarlatifkhan sorry again for the radio silence, my health has not been great. :( I'm sorry, I'm not seeing an outstanding comment to address, can you point me to it?

khawarlatifkhan commented:
@nzakas That last comment is resolved too. I think I saw an unresolved comment last time, not anymore.
imo, everything looks good now.

Thank you so much for reviewing all the comments in detail.

khawarlatifkhan commented:
@nzakas There's one unresolved comment (line 97 in the "configuration-files.md") file.

nzakas commented:
@khawarlatifkhan ah, thank you. GitHub wasn't even showing me that file because it was too big. I left a comment. Let me know if that makes sense.

khawarlatifkhan commented:
@nzakas Thank you. I've edited the content and resolved the comment.
Thank you so much for being patient and looking at all the comments in such detail. All. the comments are now resolved.

nzakas commented on docs/user-guide/configuring/configuration-files.md:
````diff
@@ -105,7 +105,7 @@ your-project
   └── test.js
 ```
 
````
Looks like we are missing the backticks around the “.eslintrc” and “root: true”

khawarlatifkhan replied:
Fixed it.

nzakas reviewed (approved):
LGTM! Thanks for all of your work. 🙏

nzakas commented:
I’m merging this because I think it’s ready. We can always make adjustments in the future if necessary, but for now let’s get this live!

btmills commented:
@khawarlatifkhan thank you, and great job on this huge improvement! 🥇 I'm excited to see it go live with the next release.

kaicataldo commented:
🎉 Amazing work @khawarlatifkhan!

khawarlatifkhan commented:
Thank you @kaicataldo @btmills @nzakas.
Your help and support means a lot.
